server/api/src/rsGenQuery.cpp

```diff
@@ -125,7 +125,84 @@ namespace {
+        static const std::regex equal_regex{R"_(^\s*=\s*$)_"};
+        static const std::regex like_regex{R"_(^\s*(LIKE|like)\s*$)_"};
+        static const std::regex not_like_regex{R"_(^\s*(NOT|not)\s+(LIKE|like)\s*$)_"};
+        static const std::regex not_equal_regex{R"_(^\s*!=\s*$)_"};
+        bool equal_op = false;
+        bool like_op = false;
+        bool not_like_op = false;
+        bool not_equal_op = false;
+
+        // clang-format off
+        if (std::regex_match(op.data(), equal_regex))          equal_op = true;
+        else if (std::regex_match(op.data(), like_regex))      like_op = true;
+        else if (std::regex_match(op.data(), not_like_regex))  not_like_op = true;
+        else if (std::regex_match(op.data(), not_equal_regex)) not_equal_op = true;
+        // clang-format on
+
+        if (!equal_op && !like_op && !not_like_op && !not_equal_op) {
+            THROW(CAT_INVALID_ARGUMENT, fmt::format("Invalid condition: [{}]", cond));
+        }
+
+        // Generate a regex by replacing GenQuery wildcards with regex wildcards.
+        std::string hier_regex_str = cond.substr(open_quote_pos + 1, close_quote_pos - open_quote_pos - 1);
+        if (!equal_op && !not_equal_op) {
+            auto wildcard_pos = hier_regex_str.find_first_of("%");
+            while (std::string::npos != wildcard_pos) {
+                hier_regex_str.replace(wildcard_pos, 1, "(.*)");
+                wildcard_pos = hier_regex_str.find_first_of("%");
+            }
+        }
+        const std::regex hier_regex{hier_regex_str};
+
+        const auto want_to_match = like_op || equal_op;
+        std::vector<rodsLong_t> leaf_ids;
+        try {
+            for (const auto& hier : resc_mgr.get_all_resc_hierarchies()) {
+                const auto match_found = std::regex_match(hier, hier_regex);
+                if ((match_found && want_to_match) || (!match_found && !want_to_match)) {
+                    leaf_ids.push_back(resc_mgr.hier_to_leaf_id(hier));
+                }
+            }
+        }
+        catch (const irods::exception& e) {
+            // Need to handle irods::exceptions here because get_all_resc_hierarchies throws when a structured file is
+            // being operated upon. Somehow, a resource with ID -1 is found in the resource map with structured files.
+            // Historically, this has not been treated as an error in this case.
+            irods::log(LOG_DEBUG, fmt::format("irods::exception occurred: [{}]", e.client_display_what()));
+            return default_condition_str;
+        }
+
+        if (leaf_ids.empty()) {
+            return default_condition_str;
+        }
+
+        // Generate IN condition with the resulting list of leaf IDs.
+        std::string condition = "IN (";
+        for (const auto& leaf_id : leaf_ids) {
+            condition += fmt::format("'{}'{}", leaf_id, leaf_id != leaf_ids.back() ? "," : ")");
+        }
+        return condition;
```

Comment:
Seems good to me

Reply:
Well, I'm not advocating to change it. I was mostly wondering which one was more readable. Use whichever one you feel is best.

It would be interesting to get some benchmarks around these. If I had to guess, your current impl may be faster, but you never know :-)

Reply:
Sure. Your solution seems better to me due to the use of the algorithmic tool and avoiding that ugly conditional :) I'll try it out and put some numbers up soon, hopefully

Reply:
You may find quick-bench useful for running simple benchmarks.

https://quick-bench.com/

Reply:
It didn't occur to me until the compiler yelled at me that this won't work because `leaf_ids` is a `std::vector<rodsLong_t>` and we are returning `std::string` in the lambda expression used in the `std::transform` call. Something like this could still work, but I'll have to tinker with it.

Reply:
Crap. That's true.

Should be simple enough to tweak though. Of course, up to you on which version you use.

Reply:
So, I ended up side-stepping the issue by changing the type of `leaf_ids` to `std::vector<std::string>` and pre-formatting the strings as they're pushed onto the vector. That way, we can just call `fmt::join` at the end on the resulting vector which already has the `rodsLong_t`s formatted with the single-quotes.